### DISCUSSION TITLE
遇到非 Bug/Feature 问题可以在这里提问

### DISCUSSION THREAD

**Discussion by LetTTGACO** [LetTTGACO/elog#21] [Q&A]
遇到非 Bug/Feature 问题可以在这里提问

**Comment by handbye** [comment]
我搭建好了hexo，这是我的地址：https://notion-hexo-ten.vercel.app/

但有个奇怪的问题是：hexo里面的文章从首页点进去可以正常访问，但是使用的单独的地址访问就会直接下载。

例如：https://notion-hexo-ten.vercel.app/2022/10/23/data-auit-system-rce

@LetTTGACO 可以帮忙看下是为什么吗，谢谢。

**Reply by LetTTGACO** [reply]
看起来像是 Hexo 或者 Vercel 的配置问题。我也第一次遇到这个问题，我尝试问了下 ChatGPT，他给出了一些检查措施，可能会对你有所帮助，我也会尝试重现此 bug并解决。如果你解决了，请告诉我，方便帮助到其他人谢谢
```
这个问题可能涉及到 Hexo 博客部署和 Vercel 的配置。让我们一步一步来看看可能的原因和解决方案。

1. Hexo 部署问题：
如果从博客首页进入文章详情正常加载，但单独访问或刷新文章详情会触发浏览器下载行为，首先检查 Hexo 的配置是否正确。确保 Hexo 生成的静态文件在 Vercel 中正确部署。
确认 Hexo 的 _config.yml 文件中的 permalink 配置是否正确。permalink 决定了文章详情的 URL 格式。例如，如果你的配置是 permalink: :year/:month/:day/:title/，那么文章详情的 URL 应该是类似于 https://yourdomain.com/2024/03/13/your-article-title/。
2. Vercel 配置问题：
检查 Vercel 中的部署设置。确保你的 Vercel 项目正确连接到了你的 GitHub 仓库，并且自动部署已经启用。
确认 Vercel 的构建和部署日志是否有错误信息。如果有错误，根据日志来排查问题。
3. 文件类型和 MIME 类型：
确认你的文章详情页面的文件类型和 MIME 类型是否正确。如果服务器返回的 MIME 类型不正确，浏览器可能会触发下载行为。确保服务器正确设置了 .html 文件的 MIME 类型。
```

**Reply by handbye** [reply]
感谢
是因为我_config.yml里的permalink和文章里写的permalink不匹配导致的，文章里写的permalink最后面少了一个`/`

**Comment by handbye** [comment]
遇到一个这种问题，就是我在notion里面的链接是这种格式的：
![image](https://github.com/LetTTGACO/elog/assets/26617058/014c5e6f-b77a-49eb-bc47-c867e7c91a73)
然后到了hexo中成这种的了：
![image](https://github.com/LetTTGACO/elog/assets/26617058/813c20a6-1892-4e4b-8d1c-899cb851b1ba)
有办法让他自动转成对应的url吗，而不是显示"bookmark"

**Reply by LetTTGACO** [reply]
暂时不行，notion 转 md 用的是第三方库，目前没有精力和时间复刻别人的库进行修正，我会记个代办。你可以先用标准的 md 超链接在 notion 上记录。

**Comment by lightyisu** [comment]
你好 非常感谢您写的elog。
遇到了这样一个问题：
在搭配vitepress使用的过程中，遇到了一个post排序问题。
使用 sorts:'dateDesc' 对notion数据库进行一个以date时间的倒序输出，在控制台我发现下载文档的输出是按顺序的。但是生成md到对应目录后顺序就被打乱了（此时开启了catalog=true)，因为vitepress是按照目录文件排序生成的博客，这样最后的目录也乱了。
这个怎么解决呢？
![image](https://github.com/LetTTGACO/elog/assets/45456944/effd7550-8504-4e48-9b56-7278b575c94c)
地址：https://github.com/lightyisu/yisupower 附有配置

**Reply by LetTTGACO** [reply]
好问题，因为我也遇到了，所以才逐渐找到了解决问题的办法。

目前没有特别完美的办法，因为 vitepress 的路由生成有两种方式（我知道的）
1. 自动根据目录生成，所有文档排列顺序就是根据目录来的。但是你文档下载到本地，文件夹的排列顺序不是很好受你控制，所以在 vitepress 上才会出现这个问题
2. 自定义目录和文档路由。[vitepress配置侧边栏](https://vitepress.dev/zh/reference/default-theme-sidebar)，vitepress 提供了自定义侧边栏的配置。如果想要实现你的想法，就得用这种方式：通过原始文档的目录信息，自己生成指定顺序的路由。这就要用到elog.cache.json 缓存文件中的 catalog 下的信息结构了。

实现想法就是：elog.cache.json 缓存文件中的 catalog 下的信息结构是你按照一定顺序下载下来的的原始文档信息，这里面的结构有文档标题、目录信息，这些信息足够用来手动生成路由。

具体实践课参考Elog 文档仓库中的配置：https://github.com/LetTTGACO/elog-docs/blob/master/docs/.vitepress/config.mts#L99

<img width="521" alt="image" src="https://github.com/LetTTGACO/elog/assets/37357188/5b8475cf-dc28-4b15-a19f-3486ae0a722c">

这里就是用来手动指定侧边栏，路由处理函数在这里：https://github.com/LetTTGACO/elog-docs/blob/master/utils/route.ts

我的 配置文件：https://github.com/LetTTGACO/elog-docs/blob/master/elog.config.notion.js
我的 elog 文档的 Notion地址：https://1874.notion.site/0aa9217e5bcc46768bdae424fddcbc28

可以参考我这个进行配置，有什么不懂的可以问，也可以发邮件：me@mail.1874.cool

**Reply by LetTTGACO** [reply]
我是按照 sort 字段来排序的
所以我在配置文件中指定的 sortAsc

**Reply by lightyisu** [reply]
好的非常感谢，周末我写博客的时候就去尝试一下有什么想法和问题我就联系您。
十分感谢 :)

**Comment by danmuking** [comment]
我有一个问题，由于我使用vuepress，我需要设置一些front matter属性，因此我打开了deploy.local.frontMatter.enable=true来让下载下来的md文件具有front matter，但是我发现只有elog中定义的一些属性能够被解析，如果我需要添加一些elog中没有定义的front matter属性，我该怎么做

**Reply by LetTTGACO** [reply]
根据你使用的写作平台，做对应的处理：https://elog.1874.cool/notion/raqyleng501h23p1

**Comment by li-wanxin** [comment]
您好，按照你的文档我成功搭建了基于elog的Hexo博客，但是在搭建语雀平台的webhooks消息推送的时候遇到了些问题。
如文档所说，我使用serverless-api项目接口使用postman自测是可以实现文章更新的，但是在语雀平台调用会失败，这个应该是如yuque-hexo项目中文档所说的调用国外Vercel 的服务可能会失败的原因。所以我这边考虑fork您serverless-api项目并且希望通过vercel赋予域名及阿里云平台对应解析的方式实现国内的调用。
这个方案是否可行？还有我fork了serverless-api项目后测试postman调用Vercel域名的接口会报204 no content 并且无success返回，所以serverless-api在fork后是否还需要更改vercel或者github其他设置？

**Reply by LetTTGACO** [reply]
总结一下你遇到的两个问题：
1.  因为语雀 webhooks 直接调用vercel 的 serverless会失败，你想用阿里云的 dns 解析做一层转发，然后试试可不可行。

说实话不确定，我没试过，可以尝试，貌似可以。

2. 你 fork 后部署到自己的 vercel serverless 后，调用不通，不知道怎么回事。

印象中没有什么特殊的设置，直接部署到自己的 vercel serverless 就可以，可以看看 vercel serverless 的文档。


当然你可以考虑别的办法。
你的目的就是想在语雀写完文档后能自动/快捷触发 github action 的接口，本质上就是调用那个 serverless 的接口。

1. 自己部署一个 node 服务到国内云服务器上，并实现以上接口，然后用语雀webhooks 调用。也可以使用国内平台的 serverless，但不推荐，太贵了且容易被恶意调用攻击。。。惨痛教训 orz
2. 根据我的使用体验来看，博客的更新不会十分频繁，只要自己找到方法能快捷调用同步接口就行，所以我自己常用以下几种手动触发方式
    2.1. 使用 [BlogPal 浏览器插件](https://elog.1874.cool/notion/blogpal)，配置好serverless 接口后手动触发。
    2.2 浏览器直接访问`https://serverless-api-elog.vercel.app/api/github?user=xxx&repo=xxx&event_type=xxx&token=xxx`即可触发，这个方法最简单，也支持跨平台调用，所以在万一需要在手机上触发，我会使用此方法。
    2.3 我本身有使用 macOS 的 Raycast 快捷启动工具，自己实现了一系列的同步脚本，本质上也是调用 2.2 的接口，能够在电脑上快速触发 【我最常用的手段】

<img width="862" alt="image" src="https://github.com/LetTTGACO/elog/assets/37357188/40592ab4-e081-40a2-8b96-a0aa3d7522c1">

**Comment by VWmin** [comment]
在用confluence，有对应的例子吗，搞不太明白

**Reply by LetTTGACO** [reply]
> 在用confluence，有对应的例子吗，搞不太明白

这个之前只有我自己在公司内部用，目前文档没那么完善。可以邮件联系我相关细节问题
me@mail.1874.cool

**Comment by ljw-kr** [comment]
elog同步下来的文章的顺序跟语雀的不一样(下载顺序不同，实际文档的排序也不同)，想问下elog是根据什么去排序的，还是说就是随机的？ 文档没看到哪里有配置的地方

**Reply by LetTTGACO** [reply]
这个需要自己进行处理的，可以参考我的[仓库配置](https://github.com/LetTTGACO/elog-docs/blob/master/utils/route.ts)

因为 vitepress 或者其他平台大多都是自动根据某个目录中的 md 文件来自动生成导航信息，而目录中文件排列顺序是首字母排序（或者其他排序），而用户本身在语雀上的排序并不是此排序，所以就导致了生成的导航顺序和语雀不一致

解决办法：
1. 既然vitepress是根据目录本身的排序来生成导航栏，那么就可以手动在语雀上变更文件名，例如：001_文件 1、002_文件 2，让生成到本地的 md 文件目录的排列顺序和语雀保持一致即可

2. vitepress 或者其他平台一般都是可以自定义导航顺序（甚至可以手动在配置文件录入导航层级信息），所以可以通过工具函数来自动根据语雀的导航层级顺序来自动生成导航信息。所以就需要知道语雀本身的目录顺序信息，此信息被保存在  elog.cache.json 中，通过工具函数读取此文件的数据，然后生成vitepress的导航顺序就行。（我采用的办法）

**Comment by zacharydigital** [comment]
有没有 WordPress +notion的最佳实践demo示例，   在notion 写作同步到WordPress，用了很多第三方插件解决方案都存在问题，block块同步完整的问题
